docs/adr/0031-process-evolution.md

```diff
@@ -0,0 +1,171 @@
+drawback we should continue to use tags.
+
+### Who uses the change logs and what are their expectations?
+
+We’ve maintained change logs in each package for many years and they are
+updated with each release. We release often enough that not all of the
+packages change between each release and we regularly update the changelogs
+just to say nothing has changed. This causes considerable churn in our
+deployment pipeline for no reason.
+
+We asked our stakeholders about the changelogs and the only people currently
+using them are QA. After [a
+conversation](https://groups.google.com/a/mozilla.com/g/fxa-staff/c/OuKEsHs67fk)
+we found an alternative that would allow us to stop writing our change logs to
+the repository and reduce our deployment churn.
+
+### Do we need to maintain versions in each package.json?
+
+Our build script currently increments the version of every package in our
+monorepo when we run it. But do we need those versions? [The docs
+say](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#version) “If
+you don't plan to publish your package, the name and version fields are
+optional.” We used to publish our packages to npm but we don’t do that anymore.
+SRE doesn't use the version field in any scripts. We use the version to [tell
+what is live](https://accounts.firefox.com/__version__) but the commit hash is
+also there and we could look at other options.
+
+So, no, we don't need the version in each package.json.
+
+## Decision Outcome
+
+TBD, but the authors are recommending option 3
+
+## Options
+
+### Option 1: The main branch is always ready to push live
+
+![An illustration of Option 1](images/0031-option1.png)
+
+Something like [Github Flow](https://docs.github.com/en/get-started/quickstart/github-flow).
+This is mainly used for continuous deployment which we aren’t aiming for so likely
+isn't a good fit here.
+
+#### Pros:
+
+- Forced robustness in DB/data schema changes
+- Never have to “roll back” a release? I could be wrong about this
+
+#### Cons:
+
+- Have to carefully juggle feature flags
+- L10n has no time to push unless we land the code early behind a feature flag
+- This would encourage feature branches which we’ve discouraged up until now to
+  avoid merging problems like conflicts
+- Doesn’t scale well with large teams
+
+### Option 2: A single long running release branch
+
+![An illustration of Option 2](images/0031-option2.png)
+
+#### Pros
+
+- Easy to understand
+
+#### Cons
+
+- We’d have to set up infrastructure somewhere else to push to production
+  making this less simple or effecient
+
+### Option 3: Long running stage and production branches
+
+![An illustration of Option 3](images/0031-option3.png)
+
+We make a new long-running production branch (essentially [GitLab Flow](https://docs.gitlab.com/ee/topics/gitlab_flow.html)). From that link:
+
+> After announcing a release branch, only add serious bug fixes to the branch. If
```

Comment:
I know this is a direct quote but when I first read "After announcing a release branch," I read it as us creating a new branch for every release (similar to our current train branches) that we then "announce" to the team or something. 🤔

I think option 3 is proposing we have `main` which is our development branch, then we have `staging`, then `prod` or `production`. I guess this means that we will always have to merge `staging` into `production` whenever SRE wants to push to prod (might be a con since that's an extra step over what we have now). I'm not quite sure how we would handle [production hotfixes](https://mozilla.github.io/ecosystem-platform/reference/team-processes/release-process#production-hotfixes) with this when we have a different major release in staging than what we have in prod, but I didn't read all of GitLab's docs.

Reply:
> 
> I'm not quite sure how we would handle [production hotfixes](https://mozilla.github.io/ecosystem-platform/reference/team-processes/release-process#production-hotfixes) with this when we have a different major release in staging than what we have in prod, but I didn't read all of GitLab's docs. 

Yeah, since we are going to continue to use tags, and (the ADR made no explicitly statements on this) if we also continue with our current release and tagging cadence, the stage branch cannot be updated in a fast-forward fashion as implied by the image.

Reply:
Yes, there are only 3 branches that we regularly use: `main`, `staging`, and `production`.  

If a production hotfix is needed, I would expect it to land on `main` and be cherry-picked to `staging` and `production`.  Altering SRE processes are fair game if needed.

@chenba can you say more about "the stage branch cannot be updated in a fast-forward fashion" ?

Reply:
> If a production hotfix is needed, I would expect it to land on main and be cherry-picked to staging and production. Altering SRE processes are fair game if needed.

Would you still be able to reliably test this hotfix, given `staging` and `production` have potentially already diverted? I assume `staging` would be in a different state for a period before release, at least for testing purposes.

Coming from a different product, although much simpler (legal-docs), merging changes from the default branch to `prod` was a nightmare, even leading to unreliable content in prod. At this point, we're just pushing content from one branch to the other, and gave up on the idea of having a shared history.

Reply:
> @chenba can you say more about "the stage branch cannot be updated in a fast-forward fashion" ?

1. Production is at tag `n`.
1. It's Wednesday and `n+1` is tagged from `main`.  `n+1` is pushed or force pushed to `staging`.
1. An issue was found and fixed on `main` with commit `p`.
1. `p` is cherry picked onto the HEAD of `production` in some "hotfix" branch, and the resulting commit, `q`, is tagged and force pushed to `staging`.  Then the new tag `n'` is pushed to `production`.
1. `p` is cherry picked onto the commit from which we tagged `n+1`, and the resulting commit, `r`, is tagged and force pushed to `staging`.

All the force pushes mean that the `staging` and `production` branches are not simply moving from one commit to the next as the directed graphs neatly illustrates.  It's not _a real problem_ if we don't care that the patch release commits of the older "major versions" are not in `staging` and `production`; we can always rely on the tags.

There's further complication with the auto-deployments on these force pushes to `staging` if `n` and `n+1` have incompatible configuration value changes, but that should be super rare.

Reply:
> > If a production hotfix is needed, I would expect it to land on main and be cherry-picked to staging and production. Altering SRE processes are fair game if needed.
> 
> Would you still be able to reliably test this hotfix, given `staging` and `production` have potentially already diverted? I assume `staging` would be in a different state for a period before release, at least for testing purposes.

I think we will have to alter at least some process if we want `staging` to always reflect staging and `production` to always reflect production since any new tag pushes the release to staging.

I like the idea of cherry-picking the fix into both branches and tagging each branch as separate patch releases, but I also wonder about @flodolo's point of being able to test it before it's just pushed to prod. I know we used to have a "dev" environment as well as staging... is it feasible to set up another environment to test production hotfixes? I wonder if we could automatically deploy tags with some flag (maybe something like `-hotfix` or `-testprod` in the tag name) from the `production` branch to it similarly to how we automatically deploy all new tags to staging now, and then when it's verified we push to production. This would be rare thankfully.

Reply:
@flodolo I would expect an emergency fix in production to be a pretty small piece of code but we always have options to test it behind a flag, or run a canary instance or something.  It'll be rare that `main` will have diverged so rapidly from `production` that we won't be able to test it anyway.  So, it'll depend on whatever situation we're in, but for the most part I wouldn't expect testing a hotfix to be a problem - we have options.

Reply:
@chenba I think you're assuming there is no change to SRE process which isn't necessarily true.  In step 4 I don't think `q` needs to be pushed to staging at all (I'm assuming you're doing that to build packages?).  I think your steps 1-3 make sense, and then:

4. `p` is cherry picked onto production and pushed.  This gets the emergency out of the way.  We can generate a new tag and release on github for this.
5. `p` is cherry picked onto stage and pushed.  We can generate a new tag with this fix, as well as a new _prerelease_ on github.

Reply:
@clouserw I think I would personally prefer that we swap steps 4 and 5, and to always verify the hotfix on staging first before tagging the production branch. That will help with the concerns around testing a fix before it's immediately live in production.

Reply:
I haven't seen this mentioned explicitly but I think @LZoog actually touched on it. Terminology aside, if we have the bandwith to run more shared environments one option could be to run : 

- production/live where we deploy the current `production` branch
- stage where we deploy only the current `staging` branch. this would be an exact replica of the production environment and the only delta between the `staging` branch and `production` branch would be the delta between what is currently on production and what would be if the build on stage is promoted (this would either be just a hotfix mid release cycle or the delta between `main` and `production` at the end of a sprint).
- dev/qa where we deploy the current `main` for qa/other stakeholders to preview what is in the works for the next release

Reply:
side note: that in the above, you could share environments as long as you know what is deployed to each

Reply:
>     * dev/qa where we deploy the current `main` for qa/other stakeholders to preview what is in the works for the next release

Can we make this explicit in the ADR?

Reply:
Thanks for the comment, @IvoJP.  I think your production and stage observations are what we do today and the dev/qa proposal would be a different discussion than this ADR about what we would want that service to look like.  In fact, [we have one already and its unused](https://github.com/mozilla/fxa/issues/9506), so any change there would be a discussion with the people who would want to use it.